runtime/runtime-params-estimator/src/gas_cost.rs

```diff
@@ -104,6 +104,18 @@ impl GasCost {
             }
         }
     }
+
+    /// Subtracts two gas costs from each other without panicking on an arithmetic underflow.
+    /// If the given tolerance is breached, the result will be marked as uncertain.
+    pub(crate) fn safe_subtract(&self, rhs: &Self, tolerance: &NonNegativeTolerance) -> Self {
```

Comment:
Let's call this `saturating_sub` maybe?

Reply:
Wouldn't that lead to confusion with the existing `impl num_traits::SaturatingSub for GasCost`?

Reply:
As an aside, I'd just make that `saturating_sub` an inherent impl: using num_traits here feels like a case of [concrete abstraction](https://matklad.github.io/2020/08/15/concrete-abstraction.html)

Reply:
Agreed on removing `num_traits` abstraction. And I think the calls to the old `saturating_sub` can be replaced with calls to the new function that merges `saturating_sub` and `safe_subtract`.

For the other change, I am not so sure. Personally, I don't like hiding important details that are not obvious in the function signature. The uncertain flag is a very important details, if it is set the entire estimation is essentially null and void.

Saturating subtraction or `ops::Sub` assuming a strict tolerance seems wrong, IO bytes as we measure it today will always have cases that underflow sometimes. Using 0.1% as default hard-coded tolerance would be ok but hard-coding such numbers smells to me.

I have two suggestions, either to swallow complexity or magic:

1. Complexity: Get rid of `ops::Sub` and embrace the fact that `GasCost` cannot be subtracted like a primitive. All subtractions should use `saturating_sub()` with a chosen tolerance.
2. Magic: `saturating_sub` has a tolerance argument and `ops::Sub` calls that with a default value of 0.1%. This way, choosing `-` has some magic hidden behind it but at least is never really wrong. Estimations functions MAY opt-in to use an explicit tolerance by calling `saturating_sub()` if their local knowledge requires it. (For example, when no tolerance is desired)

WDYT?

Reply:
>Saturating subtraction or ops::Sub assuming a strict tolerance seems wrong, I

Good point. I think there's also 3:

Make `ops::Sub` *strict* (always set uncertain on underflow), but also add `tolerance_sub` or something which allows to *not* set uncertain if it's close enough.

I don't have strong preference among 1, 2, 3 -- all seems workable and refactorabe. I suggest picking one and seeing how it goes

Reply:
I decide for 3, I like strict `-` better than some hard-coded arbitrary number.